flink-kubernetes-operator/src/main/java/org/apache/flink/kubernetes/operator/observer/AbstractFlinkResourceObserver.java

```diff
@@ -0,0 +1,129 @@
+
+        SavepointUtils.resetTriggerIfJobNotRunning(resource, eventRecorder);
+    }
+
+    /**
+     * Get the observer context for the current resource.
+     *
+     * @param resource Resource being observed
+     * @param context Resource context
+     * @return Observer context
+     */
+    protected abstract CTX getObserverContext(CR resource, Context<?> context);
+
+    /**
+     * Check whether the resource should be observed. In certain states such as suspended
+     * applications or in-progress upgrades and rollbacks, observing is not necessary.
+     *
+     * @param resource Current resource
+     * @param resourceContext Resource context
+     * @param observerContext Observer context
+     * @return True if we should observe the resource
+     */
+    protected boolean isResourceReadyToBeObserved(
+            CR resource, Context<?> resourceContext, CTX observerContext) {
+        var reconciliationStatus = resource.getStatus().getReconciliationStatus();
+
+        if (reconciliationStatus.isBeforeFirstDeployment()) {
+            logger.debug("Skipping observe before first deployment");
+            return false;
+        }
+
+        if (reconciliationStatus.getState() == ReconciliationState.ROLLING_BACK) {
+            logger.debug("Skipping observe during rollback operation");
+            return false;
+        }
+
+        // We are in the middle or possibly right after an upgrade
+        if (reconciliationStatus.getState() == ReconciliationState.UPGRADING) {
+            // We must check if the upgrade went through without the status upgrade for some reason
+            checkIfAlreadyUpgraded(resource, resourceContext, observerContext);
+            if (reconciliationStatus.getState() == ReconciliationState.UPGRADING) {
+                ReconciliationUtils.clearLastReconciledSpecIfFirstDeploy(resource);
+                logger.debug("Skipping observe before resource is deployed during upgrade");
+                return false;
+            }
+        }
+
+        return true;
+    }
+
+    /**
+     * Internal observer logic specific to each resource type.
+     *
+     * @param resource Resource to be observed
+     * @param resourceContext Resource context
+     * @param observerContext Observer context
+     */
+    protected abstract void observeInternal(
+            CR resource, Context<?> resourceContext, CTX observerContext);
+
+    /**
+     * Checks a resource that is currently in the UPGRADING state whether it was already deployed
+     * but we simply miss the status information. After comparing the target resource generation
+     * with the one from the possible deployment if they match we update the status to the already
+     * DEPLOYED state.
+     *
+     * @param resource Flink resource.
+     * @param resourceContext Context for resource.
+     * @param observerContext Context for observer.
+     */
+    protected abstract void checkIfAlreadyUpgraded(
```

Comment:
Maybe this should be called `updateStatusToDeployedIfAlreadyUpgraded` or something more meaningful. W/o checking the function description I can't really tell what is the outcome of this.